napari/viewer.py

```diff
@@ -62,6 +64,13 @@ def __init__(
             )
             raise RuntimeError(message)
 
+        # If using perfmon we need a special QApplication. If we are run from
+        # the command line using our gui_qt context object then we already have
+        # the special one, and this will be a noop. When running inside IPython
+        # or Jupyter however this is where we switch to our QApplication.
+        if perf.USE_PERFMON:
+            app = convert_app_for_timing(app)
```

Comment:
if we can do this here, do we need it in `event_loop.py` at all?  Or do you think it's still better to do it before instantiation if at all possible?

Reply:
The one in Viewer does not work for `gui_qt` because we've already created a QWidget (the splash screen) so we can't delete QApplication, it will crash. But with IPython/Jupyter we have not created any Qt objects at that point.

We could call `convert_app_for_timing` in `gui_qt`. That is call `convert_app_for_timing` in both places. But that means we'dcreate the `QApplication` then **immediately** delete it which seems kind of silly, since it's just as easy to create the right one.

So right now:
Command Line: create the right QApplication from the start.
IPython/Jupyter: delete the existing QApplication and replace it.

Reply:
oh fascinating!  So is that the distinction here?  For instance, if someone is using Jupyter and they create some other widget (outside of napari) before running napari with perfmon, will it crash there to?  (recognize that's basically never going to happen... just want to make sure I understand)

I'm definitely fine with the current approach, just wanted to understand

Reply:
I think that's right. I think napari would delete the "shared" QApplication and it would crash. If it was not the first. Running multiple napari windows should be okay. Because the 2nd one will see it's already using QApplicationWithTiming.

Anyway this worked for timing #1320 Juan's notebook so that was the immediately goal. But this could bite someone if they just kept perfmon enabled, it was be a very mysterious crash for them. Might take a long while to suspect perfmon.

Between this and the crash on exit it'd be **really** nice to find a completely new way of timing events. I'd be nice if perfmon were bulletproof. But I'm probably going to focus on #1320 itself.

Maybe there can be a V2 of perfmon at some point.

Reply:
yeah that works for me.  as an internal tool, it's plenty useful as is